refactor(InMemorySinger) - Implement a InMemorySigner::test method

mkamonMdt opened the pull request:
Following the issue #12503 the commit provides and factory methods for testing purposes only. Use case examples are also provided.

The general idea is to wrap verbose but prevalent pattern of setting up a test case with  calls.

mkamonMdt commented:
@wacban  Please check if we are on the same page with the implementation for #12503. If so I could refactor few more usages. 

Also I had a problem with `#[cfg(feature="rand"]` to ensure that methods are enforced into test code only.  he `#[cfg(feature = "rand")]` isn't very strict. It will compile the code anytime the feature = "rand" is in list of dependencies of a create as far as I understand. Ideally we could have something like `#[cfg(all(test, feature ="rand"))]`  but it will not compile with integration tests. 

Another option would be s #[cfg(any(all(test, feature = "rand"), feature = "rand-test-utils"))]  and adding dependency crate-by-crate where it is necessary.  Please let me know if there is a near-way to approach that or should I leave it as is.

wacban commented:
> Please check if we are on the same page with the implementation for https://github.com/near/nearcore/issues/12503. 

We are, thanks! 

> #[cfg(feature="rand"]

This shouldn't be necessary. What breaks without it? 

It's ok to not enforce test-only usage, it should be clear enough from the method name. If you still want to do this you can use cfg test.

wacban commented on tools/state-viewer/src/contract_accounts.rs:
```diff
@@ -721,8 +721,7 @@ mod tests {
     /// Create a test receipt from sender to receiver with the given actions.
     fn create_receipt_with_actions(sender: &str, receiver: &str, actions: Vec<Action>) -> Receipt {
         let sender_id: AccountId = sender.parse().unwrap();
-        let signer =
-            InMemorySigner::from_seed(sender_id.clone(), near_crypto::KeyType::ED25519, "seed");
```
this is different, the account and the seed are not the same. Please revert it - or make sure that it's ok. The latter would require testing this neard sub command so it's probably easier and safer just to leave it as is.

mkamonMdt replied:
fixed, thanks.

wacban reviewed (commented):
Looks, great, I love that you're reducing the line count by more than a hundred! 

I'll see what's going on with rand, it really shouldn't be needed there, I'll come back to you shortly.

wacban commented:
@mkamonMdt I removed the rand feature gating locally and the code compiles and the tests are passing. Can you remove it please? Let's see if it passes the whole suite of tests in CI.

mkamonMdt commented:
> @mkamonMdt I removed the rand feature gating locally and the code compiles and the tests are passing. Can you remove it please? Let's see if it passes the whole suite of tests in CI.

Fixed :)

wacban reviewed (approved):
LGTM, thank you!

mkamonMdt commented:
Ok it seems it needs feature="rand" after all, the "Non-default configuration Builds job failed"  :( 
![Screenshot from 2024-11-27 18-09-22](https://github.com/user-attachments/assets/63facc01-2c0d-4e07-9047-47dd98db6fa7)

wacban commented:
> Ok it seems it needs feature="rand" after all, the "Non-default configuration Builds job failed" :(

Adding it to the InMemorySigner is still the wrong thing to do, we should fix that at the source. I'll have a proper look tomorrow.

wacban commented:
Ah nevermind, the reason it breaks is because the new methods depend on the from_seed function that is already gated by rand. All good!
